hack/dind-cluster.sh

```diff
@@ -263,12 +267,161 @@ function build-images() {
 function build-image() {
   local build_root=$1
   local image_name=$2
+  local dockerfile=${3:-}
 
+  if [[ -n "${dockerfile}" ]]; then
+    dockerfile="-f ${dockerfile}"
+  fi
   pushd "${build_root}" > /dev/null
-    ${DOCKER_CMD} build -t "${image_name}" .
+    ${DOCKER_CMD} build -t "${image_name}" ${dockerfile} .
   popd > /dev/null
 }
 
+function start-ansible-cluster() {
+  local cluster_id=$1
+  local private_key=$2
+  local force_config_write=$3
+
+  echo "Starting ansible dind cluster '${cluster_id}'"
+  ssh-keygen -N '' -q -f "${private_key}" > /dev/null
+
+  ensure-host-configuration "openshift/dind-host"
+
+  local run_cmd="${DOCKER_CMD} run -d --privileged"
+
+  # Create containers
+  local master_cid; master_cid="$(${run_cmd} --name="${MASTER_NAME}" --hostname="${MASTER_NAME}" "${HOST_IMAGE}")"
+  local master_ip; master_ip="$(get-docker-ip "${master_cid}")"
+  set-authorized-keys "${MASTER_NAME}" "${private_key}"
+  local node_ips=()
+  local cid
+  for name in "${NODE_NAMES[@]}"; do
+    cid="$(${run_cmd} --name="${name}" --hostname="${name}" "${HOST_IMAGE}")"
+    node_ips+=( "$(get-docker-ip "${cid}")" )
+    set-authorized-keys "${name}" "${private_key}"
+  done
+
+  # TODO support a variable number of nodes
+
+  local inventory_file="/etc/ansible/hosts"
+  if [[ -f "${inventory_file}" && -z "${force_config_write}" ]]; then
+    echo "WARNING: ${inventory_file} already exists.  Remove or use -f to overwrite.
+master ip: ${master_ip}
+node 1 ip: ${node_ips[0]}
+node 2 ip: ${node_ips[1]}" >&2
+    return
+  fi
+
+  echo "Writing inventory file to ${inventory_file}"
+  sudo mkdir -p /etc/ansible
+  sudo bash -c "cat  >> ${inventory_file}" << EOL
+# Create an OSEv3 group that contains the masters and nodes groups
+[OSEv3:children]
+masters
+nodes
+
+# Set variables common for all OSEv3 hosts
+[OSEv3:vars]
+# SSH user, this user should allow ssh based auth without requiring a password
+ansible_ssh_user=root
+
+# If ansible_ssh_user is not root, ansible_become must be set to true
+#ansible_become=true
+
+deployment_type=origin
+
+# uncomment the following to enable htpasswd authentication; defaults to DenyAllPasswordIdentityProvider
+#openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true', 'challenge': 'true', 'kind': 'HTPasswdPasswordIdentityProvider', 'filename': '/etc/origin/master/htpasswd'}]
+
+# host group for masters
+[masters]
+${master_ip}
+
+# host group for nodes, includes region info
+[nodes]
+${master_ip} openshift_node_labels="{'region': 'infra', 'zone': 'default'}" openshift_schedulable=false
+${node_ips[0]} openshift_node_labels="{'region': 'primary', 'zone': 'east'}"
+${node_ips[1]} openshift_node_labels="{'region': 'primary', 'zone': 'west'}"
+EOL
+
+  if ! grep -q "${private_key}" ~/.ssh/config; then
+    local host_match; host_match="$(echo "${master_ip}" | sed -e 's|.[0-9]\{1,3\}$|.?|g')"
```

Comment:
You probably meant to put a `\` before the `.` on the left side (though it looks like it will end up working anyway). Also, it's weird to specify `g` since the pattern can only match once anyway.